# Lab 04- Radius of The Earth

### Lillian Hanson, 09/24/2025, GPGN318A

In [21]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from scipy.constants import foot

#### Processing Calibrations
The relative gravimeters need to be calibrated to give us actual mGal values. The CG5 already accounts for this and the number it gives you is in units of mGals. For the LNR though, you must account for the calibration by hand, we have attached the (somewhat outdated) calibration sheet for our specific LNR meter. You can use this to calculate the relative gravity from the dial reading as follows:

Denote your dial reading as A.
In the supplied table (see attachment), find the nearest integer smaller than your reading from the column labeled “Counter Reading”, denote this as B.
Find the corresponding mGal value in the column labeled as “Value in mGal”, denote this as C.
Find the corresponding factor value from the column labeled as “Factor for Interval”, denote this as F.
Calculate the gravity in mGal using the formula: $g = C + F * (A - B)$.
Note: You might notice this equation as adjusting the gravity value using a linear interpolation between two known values.

In [6]:
df_lf = pd.read_csv('data/LNR_gravimeter(Sheet1).csv')
df_lf["B"] = None
df_lf["C"]= None
df_lf["F"]= None
df_lf = df_lf.rename(columns={'Digital Reading':'A'})

In [7]:
# add in values from table


# B-Values
df_lf.loc[0, 'B'] = 3100
df_lf.loc[1, 'B'] = 3100
df_lf.loc[2, 'B'] = 3100
df_lf.loc[3, 'B'] = 3100
df_lf.loc[4, 'B'] = 3100
df_lf.loc[5, 'B'] = 3100
df_lf.loc[6, 'B'] = 3100
df_lf.loc[7, 'B'] = 3100
df_lf.loc[8, 'B'] = 3100

# C-Values
df_lf.loc[0, 'C'] = 3169.70
df_lf.loc[1, 'C'] = 3169.70
df_lf.loc[2, 'C'] = 3169.70
df_lf.loc[3, 'C'] = 3169.70
df_lf.loc[4, 'C'] = 3169.70
df_lf.loc[5, 'C'] = 3169.70
df_lf.loc[6, 'C'] = 3169.70
df_lf.loc[7, 'C'] = 3169.70
df_lf.loc[8, 'C'] = 3169.70

# F-Values
df_lf.loc[0, 'F'] = 1.02481
df_lf.loc[1, 'F'] = 1.02481
df_lf.loc[2, 'F'] = 1.02481
df_lf.loc[3, 'F'] = 1.02481
df_lf.loc[4, 'F'] = 1.02481
df_lf.loc[5, 'F'] = 1.02481
df_lf.loc[6, 'F'] = 1.02481
df_lf.loc[7, 'F'] = 1.02481
df_lf.loc[8, 'F'] = 1.02481

#lol they're all the same and i put them each in manually rip

# Add row with computed G-values
df_lf["mgal"] = 3169.70 + 1.02481 * (df_lf["A"] - 3100)

df_lf

Location         A  Times            Operator     B       C        F  \
0       1.0  3144.200  13:57       Anna Williams  3100  3169.7  1.02481   
1       1.0  3144.350  14:01          Ben Koller  3100  3169.7  1.02481   
2       1.0  3144.080  14:05  Zach Shallenberger  3100  3169.7  1.02481   
3       2.0  3142.640  14:30  Sam & Tessa & Tyna  3100  3169.7  1.02481   
4       3.0  3139.242  14:45             Krishna  3100  3169.7  1.02481   
5       3.0  3139.300  14:50                 Joe  3100  3169.7  1.02481   
6       2.0  3142.830  14:59              Anna G  3100  3169.7  1.02481   
7       2.0  3142.705  15:04               Clara  3100  3169.7  1.02481   
8       1.0  3144.335  15:14                Alex  3100  3169.7  1.02481   

          mgal  
0  3214.996602  
1  3215.150323  
2  3214.873625  
3  3213.397898  
4  3209.915594  
5  3209.975033  
6  3213.592612  
7  3213.464511  
8  3215.134951

#### Measurement Averaging
Treat each instrument separately!

Average the gravity measurements taken at each visit to the stations. We visisted the stations in the following order:

- Station 1 : visit 1 
- Station 2 : visit 1
- Station 3 : visit 1
- Station 2 : visit 2
- Station 1 : visit 2
So you'll need to average all of the LNR readings at Station 1 : visit 1 together to obtain a single measurement for that location and time. Repeat with each of the visits to each station for each instrument (treat the two CG-5's separately).

At the end should have a single measurement for each of the three instruments at every visit: 5 for the LNR, 5 for CG-5 1 and 5 for CG-5 2.

#### Load in CG5 Data 

In [8]:
column_names = [
    "LINE","STATION","ALT","GRAV","SD","TILTX","TILTY","TEMP","TIDE",
    "DUR","REJ","TIME","DEC_TIME_DATE","TERRAIN","DATE"
]

df_cg51 = pd.read_csv('data/CG5-1.TXT', sep=r'\s+', skiprows=34, header=None, names=column_names)
print(df_cg51)

    LINE  STATION      ALT      GRAV     SD  TILTX  TILTY  TEMP   TIDE  DUR  \
0    1.0      1.0  29.1014  3621.271  0.426    8.9   -4.5 -1.10  0.016   30   
1    1.0      1.0  30.0779  3621.424  0.459   -5.3   -5.9 -1.11  0.023   30   
2    1.0      1.0  30.0779  3621.361  0.135   -2.9   -7.6 -1.15  0.026   30   
3    1.0      1.0  29.8338  3621.371  0.791   -4.6   -1.6 -1.18  0.030   30   
4    1.0      2.0  28.8572  3619.895  0.217   -3.0   -4.0 -1.14  0.040   30   
5    1.0      2.0  28.3689  3619.918  0.373    2.7   -0.7 -1.13  0.043   30   
6    1.0      2.0  27.8807  3619.847  0.226   -3.3   -3.5 -1.15  0.046   30   
7    1.0      3.0  27.3924  3616.459  0.242    0.4   -1.9 -1.18  0.053   30   
8    1.0      3.0  27.3924  3616.465  0.193    2.8   -9.7 -1.18  0.055   30   
9    1.0      3.0  27.3924  3616.481  0.287    1.2   -0.6 -1.19  0.058   30   
10   1.0      2.0  27.1482  3619.872  0.138   -6.0    0.6 -1.19  0.064   30   
11   1.0      2.0  26.9041  3619.902  0.161    4.1  

In [9]:
df_cg52 = pd.read_csv('data/CG5-2.TXT', sep=r'\s+', skiprows=34, header=None, names=column_names)
#print(df_cg52)
print(df_cg52.columns)

Index(['LINE', 'STATION', 'ALT', 'GRAV', 'SD', 'TILTX', 'TILTY', 'TEMP',
       'TIDE', 'DUR', 'REJ', 'TIME', 'DEC_TIME_DATE', 'TERRAIN', 'DATE'],
      dtype='object')


#### Assign variables to loaded values

In [11]:
# Note these are currently just placeholders for whatever values you obtain
lnr_grav = np.array(df_lf['mgal'].values)
lnr_time = np.array(df_lf['Times'].values)

times_dec = []
for time in df_lf['Times']:
    time = str(time)
    decimal_time = float(time.split(':')[0])+ float(time.split(':')[1])/60
    times_dec.append(float(decimal_time))

lnr_time = times_dec
print(lnr_time)

cg5_1_grav = np.array(df_cg51['GRAV'].values)
cg5_1_time = np.array(df_cg51['DEC_TIME_DATE'].values)

cg5_2_grav = np.array(df_cg52['GRAV'].values)
cg5_2_time = np.array(df_cg52['DEC_TIME_DATE'].values)

[13.95, 14.016666666666667, 14.083333333333334, 14.5, 14.75, 14.833333333333334, 14.983333333333333, 15.066666666666666, 15.233333333333333]


### Question 1
Comment on the variability of the measurements at each visit for each instrument.

TEXT HERE

#### Drift Correction
There are many things that could change the actual acceleration of gravity throughout the time of our survey. To correct for this, we assume that the change is linear in time and repeated the measurement at station 1 at the end of the survey (we made a loop).

Create a function that linearly interpolates the gravity values at Station 1 as a function of time:

In [13]:
def grav_interp_new(grav_1, grav_2, time_1, time_2, time):
    time = np.array(time, dtype=float)  # ensure numeric array
    slope = (grav_2 - grav_1) / (time_2 - time_1)
    new_grav = grav_1 + slope * (time - time_1)
    return new_grav

You will need to calculate the value of gravity at station 1 at each time for each of the instruments. Then subtract that value from every measurement. If done correctly the repeated values of station 1 at the start and end should be the same and equal to 0.

In [14]:
lnr_drift_corr = lnr_grav - grav_interp_new(lnr_grav[0], lnr_grav[-1], lnr_time[0], lnr_time[-1], lnr_time)
assert lnr_drift_corr[0] == lnr_drift_corr[-1] == 0

cg51_drift_corr = cg5_1_grav - grav_interp_new(cg5_1_grav[0], cg5_1_grav[-1], cg5_1_time[0], cg5_1_time[-1], cg5_1_time)
cg52_drift_corr = cg5_2_grav - grav_interp_new(cg5_2_grav[0], cg5_2_grav[-1], cg5_2_time[0], cg5_2_time[-1], cg5_2_time)

cg51_drift_corr = pd.Series(cg51_drift_corr, name='DRIFT_CORR')
df_cg51 = pd.concat([df_cg51, cg51_drift_corr], axis=1)

cg52_drift_corr = pd.Series(cg52_drift_corr, name='DRIFT_CORR')
df_cg52 = pd.concat([df_cg52, cg52_drift_corr], axis=1)

lnr_drift_corr = pd.Series(lnr_drift_corr, name='DRIFT_CORR')
df_lnr = pd.concat([df_lf, lnr_drift_corr], axis=1)


### Question 2
We repeated station 2 twice as well, how close are the values at this station after drift corrections?

#### Relative Gravity at each station
For each instrument you should now have their drift corrected relative gravity measurements. Since you have subtract the value of gravity of Station 1 from every Station, your measurements are relative to Station 1.

Next, for each instrument average any values at repeated stations to obtain a single measurement at each station (independant of time).

In [15]:
def drift_per_station(df):
    grav = df['GRAV'].values
    time = df['DEC_TIME_DATE'].values
    df['DRIFT_CORR'] = grav - grav_interp_new(grav[0], grav[-1], time[0], time[-1], time)
    return df
##########################################################################################
df_cg51[['STATION','GRAV','DRIFT_CORR']]
df_cg51 = df_cg51.loc[:, ~df_cg51.columns.duplicated()]

cg51_1_avg = df_cg51.loc[df_cg51['STATION'] == 1, 'DRIFT_CORR'].mean()
print(f'Relative Grav CG51, Station 1: {cg51_1_avg}')

cg51_2_avg = df_cg51.loc[df_cg51['STATION'] == 2, 'DRIFT_CORR'].mean()
print(f'Relative Grav CG51, Station 2: {cg51_2_avg}')

cg51_3_avg = df_cg51.loc[df_cg51['STATION'] == 3, 'DRIFT_CORR'].mean()
print(f'Relative Grav CG51, Station 3: {cg51_3_avg}')
############################################################################################
print('---------------------------------------')
df_cg52[['STATION','GRAV','DRIFT_CORR']]
df_cg52 = df_cg52.loc[:, ~df_cg52.columns.duplicated()]

cg52_1_avg = df_cg52.loc[df_cg51['STATION'] == 1, 'DRIFT_CORR'].mean()
print(f'Relative Grav CG52, Station 1: {cg52_1_avg}')

cg52_2_avg = df_cg52.loc[df_cg51['STATION'] == 2, 'DRIFT_CORR'].mean()
print(f'Relative Grav CG52, Station 2: {cg52_2_avg}')

cg52_3_avg = df_cg52.loc[df_cg51['STATION'] == 3, 'DRIFT_CORR'].mean()
print(f'Relative Grav CG52, Station 3: {cg52_3_avg}')
############################################################################################
print('---------------------------------------')
df_lnr[['Location','mgal','DRIFT_CORR']]
df_lnr = df_lnr.loc[:, ~df_lnr.columns.duplicated()]

lnr_1_avg = df_lnr.loc[df_lnr['Location'] == 1, 'DRIFT_CORR'].mean()
print(f'Relative Grav lnr, Station 1: {lnr_1_avg}')

lnr_2_avg = df_lnr.loc[df_lnr['Location'] == 2, 'DRIFT_CORR'].mean()
print(f'Relative Grav lnr, Station 2: {lnr_2_avg}')

lnr_3_avg = df_lnr.loc[df_lnr['Location'] == 3, 'DRIFT_CORR'].mean()
print(f'Relative Grav lnr, Station 3: {lnr_3_avg}')


Relative Grav CG51, Station 1: 0.04431880548970314
Relative Grav CG51, Station 2: -1.4660593220346527
Relative Grav CG51, Station 3: -4.89203013182896
---------------------------------------
Relative Grav CG52, Station 1: -0.025313247228106257
Relative Grav CG52, Station 2: -1.507427815110835
Relative Grav CG52, Station 3: -4.946531969048313
---------------------------------------
Relative Grav lnr, Station 1: 0.0022958405843382934
Relative Grav lnr, Station 2: -1.6086189694806308
Relative Grav lnr, Station 3: -5.1420241026621625


In [16]:
# These arrays should end up with your final relative
# values at each of the three stations for each instrument
lnr_rel = np.array([lnr_1_avg, lnr_2_avg, lnr_3_avg])
cg5_1_rel = np.array([cg51_1_avg, cg51_2_avg, cg51_3_avg])
cg5_2_rel = np.array([cg52_1_avg, cg52_2_avg, cg52_3_avg])

#### Relative to Absolute
Both the CG5 and the LNR are relative gravimeters, meaning they do not measure the full gravity value, instead they measure differences in gravity between two stations. The units of their data are in mGal (after calibration), the difference between 1 mGal and another is accurate, but a single value itself is meaningless.

Ex. Imagine someone telling you it was 1.4233 degrees celsius warmer inside than it was outside. They obtained this number by using an instrument that read out 331$^\circ$C and 332.4233$^\circ$C. The value itself doesn’t mean anything, but the difference between them is accurate. If the true temperature inside was 24$^\circ$C, then you could use this hypothetical relative temperature sensor to determine that the temperature outside was actually 24.4233$^\circ$C.

Because we know for certain the gravity value at one of the stations, we can use its value, and the relative differences from the gravimeters, to determine the absolute gravity at each of the other stations.

After drift corrections your data are explicitly the differences between each visit and station 1 ($\Delta S_i$). Use these values to calculate the absolute gravity at each station.
$$\cdot g_{abs, S_2}= \Delta (S_2)+g_{abs, S_1}$$


In [23]:
lnr_1_abs = lnr_rel[0]
lnr_2_abs = lnr_rel[1] + lnr_rel[0]
lnr_3_abs = lnr_rel[2] + lnr_rel[0]
#################################################
cg51_1_abs = cg5_1_rel[0]
cg51_2_abs = cg5_1_rel[1] + cg5_1_rel[0]
cg51_3_abs = cg5_1_rel[2] + cg5_1_rel[0]
#################################################
cg52_1_abs = cg5_2_rel[0]
cg52_2_abs = cg5_2_rel[1] + cg5_2_rel[0]
cg52_3_abs = cg5_2_rel[2] + cg5_2_rel[0]

### Question 3
For each instrument calculate an approximate gradient between stations 2 and 3.
$$\partial_zg \approx \frac{gs_3-gs_2}{\Delta H}$$ How close is this to the theoretical free air gradient of -0.3086 mGal/m? Comment on any sources of error you might have encountered.

 $\Delta H$ = 40 m

In [34]:
H_ft = 40
H = 40 * foot

grad_cg51 = (cg51_3_abs - cg51_2_abs) / H
grad_cg52 = (cg52_3_abs - cg52_2_abs) / H
grad_lnr = (lnr_3_abs - lnr_2_abs) / H

print(f'LNR Gradient: {grad_lnr:.4f} mGal/m')
print(f'CG51 Gradient: {grad_cg51:.4f} mGal/m')
print(f'CG52 Gradient: {grad_cg52:.4f} mGal/m')
print('-----------------')
print(f'LNR difference: {grad_lnr-(-0.3086):.4f} mGal/m')
print(f'CG51 difference: {grad_cg51-(-0.3086):.4f} mGal/m')
print(f'CG52 difference: {grad_cg52-(-0.3086):.4f} mGal/m')

LNR Gradient: -0.2898 mGal/m
CG51 Gradient: -0.2810 mGal/m
CG52 Gradient: -0.2821 mGal/m
-----------------
LNR difference: 0.0188 mGal/m
CG51 difference: 0.0276 mGal/m
CG52 difference: 0.0265 mGal/m


These values are fairly close to the theoretical free air gradient, $-0.3086 \space mGal/m$, with all calculated values within $0.027 \space mGal/m$ of the free air gradient. Sources of error could include:
- Vibrations near the instrument( people walking, cars moving, etc.)
- Human error in adjusting the LNR gravimeter
- Inaccurate record of time or height

## Determining the radius of the Earth
With the absolute gravity values at each location we can now find the radius of the Earth, without needing to know the mass (or the gravitational constant)! We are going to use a first order Taylor series expansion of Newton’s Universal Law of Gravitation:

### Question 4
Find an expression for the radius of the Earth at Station 2.

First, find the derivative of Newton's law with respect to $R$
$$\frac{\partial g_{abs}}{\partial R} = ...$$
Then simplify this using $g_{abs}(R)$ itself. i.e.
$$\frac{\partial g_{abs}}{\partial R}=g_{abs}(R)...$$
Use Newton's law in the Taylor series with $f(a)=g_{abs}(S_2)$ and $f(x)=g_{abs}(S_3)$ including only the first order in $x$ to show that
$$g_{abs}(S_3) \approx g_{abs}(S_2)-2\frac{g(S_2)}{R_2}\Delta H$$
Finally solve for $R_2$ to obtain an expression for the radius of the Earth at Station 2. 

### Question 5
Use your equation to obtain estimates of the Earth for each instrument. Comment on how close they were to each other and how close they were to the distance to the center of the Earth in Golden. 

### Deliverables:
- Describe the steps necessary to take a reading with each gravimeter
- Answer the questions throughout the report (5 total)
- Present the results of your attempts to measure the radius of the Earth.
- Attatch your field notebook to your report.